driver-core/src/main/com/mongodb/internal/connection/SaslAuthenticator.java

```diff
@@ -66,6 +66,11 @@ public Void run() {
 
                         res = sendSaslContinue(conversationId, response, connection);
                     }
+                    if (!saslClient.isComplete()) {
```

Comment:
From Spencer:

> But, when a server does send {done: 1}, it may send a payload which the client needs to process in order for itself to complete.

The driver is still not handling this case.  If this happens, we want to the driver to evaluate the challenge.  I think this would work:

    if (!saslClient.isComplete() {
        saslClient.evaluateChallenge((res.getBinary("payload")).getData());
        if (!saslClient.isComplete()) {
            throw ...
     }

There might be a more elegant way to code this, but I think it's correct.

Reply:
Especially since the server currently won't execute either of these branches, you'll have to fashion unit test for all three cases: 1. false on first conditional, 2. true on first, false on second, 3. true on first, true on second

Reply:
Maybe this is DRY-er:

                    while (true) {
                        response = saslClient.evaluateChallenge((res.getBinary("payload")).getData());

                        if (res.getBoolean("done", BsonBoolean.FALSE).getValue()) {
                            if (!saslClient.isComplete()) {
                                throw new MongoSecurityException();
                            }
                            break;
                        }

Reply:
Hmm.. I think this is going to break `ScramShaAuthenticator.ScramShaSaslClient#evaluateChallenge`, which throws if step > 2.

Yuck.  So maybe first suggestion is better, since it protects the call to evaluate the challenge with a check for isComplete.

Reply:
I went with the first suggestion. I added new tests, but only for the case where the outer if-conditional is true. The case where the outer if-conditional is false signals successful authentication (done=true and saslClient.isComplete()=true).